### PR TITLE
Decompile padmgr.c

### PR DESCRIPTION
**Pull request by Thar0**
🎮 🖱️ 🎤

### CODE REVIEW

File: include/padmgr.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+typedef enum {
+    /* 0 */ PADMGR_CONT_NONE,
+    /* 1 */ PADMGR_CONT_NORMAL,
+    /* 3 */ PADMGR_CONT_MOUSE = 3,
+    /* 4 */ PADMGR_CONT_VOICE_PLUGGED, // VRU plugged but not initialized
+    /* 5 */ PADMGR_CONT_VOICE,
+
+    /*-1 */ PADMGR_CONT_UNK = 0xFF
+} PadMgrCtrlrId;
```
Comment (AngheloAlf) [inline include/padmgr.h]:
```suggestion
typedef enum {
    /*  0 */ PADMGR_CONT_NONE,
    /*  1 */ PADMGR_CONT_NORMAL,
    /*  3 */ PADMGR_CONT_MOUSE = 3,
    /*  4 */ PADMGR_CONT_VOICE_PLUGGED, // VRU plugged but not initialized
    /*  5 */ PADMGR_CONT_VOICE,
    /* -1 */ PADMGR_CONT_UNK = 0xFF
} PadMgrCtrlrId;
```

File: src/code/padmgr.c [hunk 2]
```diff
@@ -1,58 +1,788 @@
...
+    for (i = 0; i < MAXCONTROLLERS; i++) {
+        if (sPadMgrInstance->ctrlrType[i] == PADMGR_CONT_NORMAL) {
+            // Check status for whether a controller pak is connected
+            if (sPadMgrInstance->padStatus[i].status & CONT_CARD_ON) {
+                if (sPadMgrInstance->pakType[i] == CONT_PAK_RUMBLE) {
+                    if (sPadMgrInstance->rumbleEnable[i]) {
+                        if (sPadMgrInstance->rumbleTimer[i] < 3) {
+                            // This should be the osMotorStart macro, however the temporary variable motorStart is
+                            // currently required for matching
+                            if (__osMotorAccess(&sPadMgrInstance->rumblePfs[i], motorStart) != 0) {
```
Comment (AngheloAlf) [inline src/code/padmgr.c:222]: despair

File: src/code/padmgr.c [hunk 3]
```diff
@@ -1,58 +1,788 @@
...
+            switch (pad->errno) {
+                case 0:
+                    // No error, copy inputs
+                    input->cur = *pad;
+                    break;
+                case (CHNL_ERR_OVERRUN >> 4):
+                    // Overrun error, reuse previous inputs
+                    input->cur = input->prev;
+                    break;
+                case (CHNL_ERR_NORESP >> 4):
```
Comment (AngheloAlf) [inline src/code/padmgr.c:456]: Personally I prefer the `case` without the extra parenthesis 
```suggestion
                case CHNL_ERR_OVERRUN >> 4:
                    // Overrun error, reuse previous inputs
                    input->cur = input->prev;
                    break;
                case CHNL_ERR_NORESP >> 4:
```

File: src/code/padmgr.c [hunk 4]
```diff
@@ -1,58 +1,788 @@
...
+
+        input->prev = input->cur;
+        isStandardController = sPadMgrInstance->ctrlrType[i] == PADMGR_CONT_NORMAL;
+        if (isStandardController) {
+            switch (pad->errno) {
+                case 0:
+                    // No error, copy inputs
+                    input->cur = *pad;
+                    break;
+                case (CHNL_ERR_OVERRUN >> 4):
```
Comment (AngheloAlf) [inline src/code/padmgr.c:452]: Could you add a new line between cases?
```suggestion
                case 0:
                    // No error, copy inputs
                    input->cur = *pad;
                    break;

                case (CHNL_ERR_OVERRUN >> 4):
```

File: src/code/padmgr.c [hunk 5]
```diff
@@ -1,58 +1,788 @@
...
+            input->cur.errno = pad->errno;
+        }
+
+        // If opposed directions on the D-Pad are pressed at the same time, mask both out
+        if ((input->cur.button & (BTN_DDOWN | BTN_DUP)) == (BTN_DDOWN | BTN_DUP)) {
+            input->cur.button &= ~(BTN_DDOWN | BTN_DUP);
+        }
+        if ((input->cur.button & (BTN_DRIGHT | BTN_DLEFT)) == (BTN_DRIGHT | BTN_DLEFT)) {
+            input->cur.button &= ~(BTN_DRIGHT | BTN_DLEFT);
+        }
```
Comment (AngheloAlf) [inline src/code/padmgr.c:487]: I wonder why it does this, since those buttons aren't used in game

File: src/code/padmgr.c [hunk 6]
```diff
@@ -1,58 +1,788 @@
...
+                sPadMgrInstance->ctrlrType[i] = PADMGR_CONT_VOICE;
+                sVoiceInitStatus = VOICE_INIT_SUCCESS;
+                func_801A4EB0();
+            }
+        }
+    }
+    // If sVoiceInitStatus is still 1 after the first attempt to initialize a VRU, don't try again
+    if (sVoiceInitStatus == VOICE_INIT_TRY) {
+        sVoiceInitStatus = VOICE_INIT_FAILED;
+    }
```
Comment (AngheloAlf) [inline src/code/padmgr.c:530]:
```suggestion
    // If sVoiceInitStatus is still VOICE_INIT_TRY after the first attempt to initialize a VRU, don't try again
    if (sVoiceInitStatus == VOICE_INIT_TRY) {
        sVoiceInitStatus = VOICE_INIT_FAILED;
    }
```

File: src/code/padmgr.c [hunk 7]
```diff
@@ -1,58 +1,788 @@
...
+ *
+ * NOTE: This function does not lock the inputs while reading them.
+ * The padmgr thread may interrupt this function and poll new inputs while the thread running this function is
+ * currently copying the old inputs potentially resulting in a mix of newly polled and old inputs being used.
+ * It is preferable to use `PadMgr_GetInput` or `PadMgr_GetInput2`.
+ *
+ * @param inputs   Array of Input of length MAXCONTROLLERS to copy inputs into
+ * @param gamePoll True if polling inputs for updating the game state
+ */
+void PadMgr_GetInputNoLock(Input* inputs, s32 gameRequest) {
```
Comment (AngheloAlf) [inline src/code/padmgr.c:672]:
```suggestion
 * @param gameRequest True if polling inputs for updating the game state
 */
void PadMgr_GetInputNoLock(Input* inputs, s32 gameRequest) {
```

**Review by AngheloAlf — COMMENTED** [review]
Great work!

Could you add the renamed functions to `namefixer.py`?

**Review by AngheloAlf — APPROVED** [review]

File: src/code/padmgr.c (hunk 2 above)
Comment (hensldm) [reply src/code/padmgr.c:222]: Based on ultralib, `__osMotorAccess`'s second arg should be `s32` which lets you remove the temp.

File: src/code/padmgr.c [hunk 8]
```diff
@@ -1,58 +1,797 @@
...
+        newY = 0.0f;
+    }
+
+    magnitude = sqrtf(SQ(newX) + SQ(newY));
+    if (magnitude > 1.0f) {
+        magnitude = 1.0f;
+    }
+    angle = Math_Atan2F_XY(curY, -curX);
+    newX = -sinf(angle) * magnitude;
+    newY = cosf(angle) * magnitude;
```
Comment (hensldm) [inline src/code/padmgr.c:426]:
```suggestion
    if(1) {
        s32 pad;

        magnitude = sqrtf(SQ(newX) + SQ(newY));
        if (magnitude > 1.0f) {
            magnitude = 1.0f;
        }
        angle = Math_Atan2F_XY(curY, -curX);
        newX = -sinf(angle) * magnitude;
        newY = cosf(angle) * magnitude;
    }
```
This lets you remove the gotos, and seems reasonable based on debug. Couldn't use magnitude and angle for some reason as the scoped variable.

File: tools/disasm/functions.txt [hunk 9]
```diff
@@ -3164,33 +3164,33 @@
...
-    0x801757A4:("Padmgr_ParseState",),
-    0x801759BC:("func_801759BC",),
-    0x80175AE0:("func_80175AE0",),
-    0x80175C80:("Padmgr_Update",),
-    0x80175E3C:("Padmgr_Stop",),
-    0x80175E68:("func_80175E68",),
-    0x80175F98:("Padmgr_GetInput",),
-    0x80175FD4:("Padmgr_GetInput2",),
-    0x80176010:("Padmgr_ThreadEntry",),
+    0x801754E0:("PadMgr_CalcStickEdges",),
```
Comment (hensldm) [inline tools/disasm/functions.txt:3184]:
```suggestion
    0x801754E0:("PadMgr_AdjustInput",),
```

**Review by hensldm — APPROVED** [review]

File: src/code/padmgr.c [hunk 10]
```diff
@@ -1,58 +1,797 @@
...
+/**
+ * Unsets the rumble retrace callback if the provided function and argument are the same as those already registered.
+ *
+ * @param callback the callback to test against
+ * @param arg the argument to test against
+ *
+ * @see PadMgr_SetRumbleRetraceCallback
+ */
+void PadMgr_UnsetRumbleRetraceCallback(void (*callback)(void*), void* arg) {
+    if (sPadMgrInstance->rumbleRetraceCallback == callback && sPadMgrInstance->rumbleRetraceArg == arg) {
```
Comment (engineer124) [inline src/code/padmgr.c:86]:
```suggestion
    if ((sPadMgrInstance->rumbleRetraceCallback == callback) && (sPadMgrInstance->rumbleRetraceArg == arg)) {
```
IIRC I think your preference is without the brackets, but we've been trying to enforce a consistent style of using brackets in MM

File: src/code/padmgr.c [hunk 11]
```diff
@@ -1,58 +1,797 @@
...
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_GetControllerBitmask.s")
+#define PADMGR_RETRACE_MSG (1 << 0)
+#define PADMGR_PRE_NMI_MSG (1 << 1)
+#define PADMGR_NMI_MSG (1 << 2)
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/func_80174F24.s")
+typedef enum {
+    VOICE_INIT_FAILED, // voice initialization failed
+    VOICE_INIT_TRY,    // try to initialize voice
+    VOICE_INIT_SUCCESS // voice initialized
```
Comment (engineer124) [inline src/code/padmgr.c:44]:
```suggestion
    /* 0 */ VOICE_INIT_FAILED, // voice initialization failed
    /* 1 */ VOICE_INIT_TRY,    // try to initialize voice
    /* 2 */ VOICE_INIT_SUCCESS // voice initialized
```

File: src/code/padmgr.c [hunk 12]
```diff
@@ -1,58 +1,797 @@
...
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_ThreadEntry.s")
+/**
+ * Immediately stops rumble on all controllers
+ */
+void PadMgr_RumbleStop(void) {
+    OSMesgQueue* serialEventQueue = PadMgr_AcquireSerialEventQueue();
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/PadMgr_Init.s")
+    for (i = 0; i != MAXCONTROLLERS; i++) {
```
Comment (engineer124) [inline src/code/padmgr.c:288]:
```suggestion
    for (i = 0; i < MAXCONTROLLERS; i++) {
```
Does this still match?

File: src/code/padmgr.c [hunk 13]
```diff
@@ -1,58 +1,797 @@
...
+                    break;
+
+                case OS_SC_PRE_NMI_MSG:
+                    actionBits |= PADMGR_PRE_NMI_MSG;
+                    break;
+
+                case OS_SC_NMI_MSG:
+                    actionBits |= PADMGR_NMI_MSG;
+                    break;
+            }
```
Comment (engineer124) [inline src/code/padmgr.c:760]: Add a default case?

**Review by engineer124 — COMMENTED** [review]
Just some minor cleanup suggestions

File: src/code/padmgr.c (hunk 13 above)
Comment (Thar0) [reply src/code/padmgr.c:760]: I don't think an empty default case is any more clear than omitting it, at least in this case

Comment (AngheloAlf) [reply src/code/padmgr.c:760]: Not having an empty `default` case should be fine, yeah

File: include/io/controller.h [hunk 14]
```diff
@@ -81,13 +123,34 @@
 #define CONT_ADDR_CRC_ER        0x04
 #define CONT_EEPROM_BUSY        0x80
 
+// Accesory detection
```
Comment (engineer124) [inline include/io/controller.h:126]:
```suggestion
// Accessory detection
```

**Review by engineer124 — APPROVED** [review]

**Pull request merged**